File: apps/api/app/utils/providers.py
"""Provider detection utilities"""
import re
from urllib.parse import urlparse
from typing import Optional, Dict
from enum import Enum


class Provider(str, Enum):
    GREENHOUSE = "GREENHOUSE"
    LEVER = "LEVER"
    ASHBY = "ASHBY"
    SMARTRECRUITERS = "SMARTRECRUITERS"
    WORKDAY = "WORKDAY"
    ORACLE_CX = "ORACLE_CX"
    AVATURE = "AVATURE"
    SUCCESSFACTORS = "SUCCESSFACTORS"
    TALEO = "TALEO"
    ICIMS = "ICIMS"
    PHENOM = "PHENOM"
    UNKNOWN = "UNKNOWN"


PROVIDER_PATTERNS = [
    {
        "provider": Provider.GREENHOUSE,
        "patterns": [
            re.compile(r".*greenhouse\.io.*jobs.*", re.IGNORECASE),
        ],
    },
    {
        "provider": Provider.LEVER,
        "patterns": [
            re.compile(r".*lever\.co.*", re.IGNORECASE),
        ],
    },
    {
        "provider": Provider.ASHBY,
        "patterns": [
            re.compile(r".*ashbyhq\.com.*", re.IGNORECASE),
        ],
    },
    {
        "provider": Provider.SMARTRECRUITERS,
        "patterns": [
            re.compile(r".*smartrecruiters\.com.*", re.IGNORECASE),
        ],
    },
    {
        "provider": Provider.WORKDAY,
        "patterns": [
            re.compile(r".*\.wd\d+\.myworkdayjobs\.com.*", re.IGNORECASE),
        ],
    },
    {
        "provider": Provider.ORACLE_CX,
        "patterns": [
            re.compile(r".*\.fa.*\.oraclecloud\.com.*CandidateExperience.*", re.IGNORECASE),
        ],
    },
    {
        "provider": Provider.AVATURE,
        "patterns": [
            re.compile(r".*\.avature\.net.*", re.IGNORECASE),
        ],
    },
    {
        "provider": Provider.SUCCESSFACTORS,
        "patterns": [
            re.compile(r".*successfactors\.com.*", re.IGNORECASE),
        ],
    },
    {
        "provider": Provider.TALEO,
        "patterns": [
            re.compile(r".*taleo\.net.*", re.IGNORECASE),
        ],
    },
    {
        "provider": Provider.ICIMS,
        "patterns": [
            re.compile(r".*icims\.com.*", re.IGNORECASE),
        ],
    },
    {
        "provider": Provider.PHENOM,
        "patterns": [
            re.compile(r".*phenompeople\.com.*", re.IGNORECASE),
        ],
    },
]


def detect_provider(url_string: str) -> Provider:
    """Detect provider from URL"""
    try:
        url = urlparse(url_string)
        full_path = url.netloc + url.path

        for pattern_config in PROVIDER_PATTERNS:
            for pattern in pattern_config["patterns"]:
                if pattern.search(full_path):
                    return pattern_config["provider"]

        return Provider.UNKNOWN
    except Exception:
        return Provider.UNKNOWN


def extract_greenhouse_ids(url_string: str) -> Optional[Dict[str, str]]:
    """Extract Greenhouse board and job ID from URL"""
    match = re.search(r"greenhouse\.io/([^/]+)/jobs/(\d+)", url_string, re.IGNORECASE)
    if match:
        return {"board": match.group(1), "jobId": match.group(2)}
    return None


def extract_lever_ids(url_string: str) -> Optional[Dict[str, str]]:
    """Extract Lever account and posting ID from URL"""
    match = re.search(r"jobs\.lever\.co/([^/]+)/([^/]+)", url_string, re.IGNORECASE)
    if match:
        return {"account": match.group(1), "postingId": match.group(2)}
    return None


def extract_ashby_ids(url_string: str) -> Optional[Dict[str, str]]:
    """Extract Ashby company and job ID from URL"""
    match = re.search(r"jobs\.ashbyhq\.com/([^/]+)/([^/]+)", url_string, re.IGNORECASE)
    if match:
        return {"company": match.group(1), "jobId": match.group(2)}
    return None


def extract_smartrecruiters_ids(url_string: str) -> Optional[Dict[str, str]]:
    """Extract SmartRecruiters company identifier and posting ID from URL"""
    match = re.search(r"jobs\.smartrecruiters\.com/([^/]+)/([^/]+)", url_string, re.IGNORECASE)
    if match:
        return {"companyIdentifier": match.group(1), "postingId": match.group(2)}
    return None

File: packages/shared/src/providers.ts
import { Provider } from './schemas/jobPosting';

export interface ProviderPattern {
  provider: Provider;
  patterns: Array<RegExp | ((url: URL) => boolean)>;
}

export const PROVIDER_PATTERNS: ProviderPattern[] = [
  {
    provider: 'GREENHOUSE',
    patterns: [
      /^job-boards\.greenhouse\.io\/[^/]+\/jobs\/\d+/i,
      /^boards\.greenhouse\.io\/[^/]+\/jobs\/\d+/i,
      /^.*greenhouse\.io.*jobs.*/i,
    ],
  },
  {
    provider: 'LEVER',
    patterns: [
      /^jobs\.lever\.co\/[^/]+\/[^/]+/i,
      /^.*lever\.co.*/i,
    ],
  },
  {
    provider: 'ASHBY',
    patterns: [
      /^jobs\.ashbyhq\.com\/[^/]+\/[^/]+/i,
      /^.*ashbyhq\.com.*/i,
    ],
  },
  {
    provider: 'SMARTRECRUITERS',
    patterns: [
      /^jobs\.smartrecruiters\.com\/[^/]+\/[^/]+/i,
      /^.*smartrecruiters\.com.*/i,
    ],
  },
  {
    provider: 'WORKDAY',
    patterns: [
      /^.*\.wd\d+\.myworkdayjobs\.com\/.+/i,
      /^.*myworkdayjobs\.com.*/i,
    ],
  },
  {
    provider: 'ORACLE_CX',
    patterns: [
      /^.*\.fa.*\.oraclecloud\.com\/hcmUI\/CandidateExperience.*/i,
      /^.*oraclecloud\.com.*CandidateExperience.*/i,
    ],
  },
  {
    provider: 'AVATURE',
    patterns: [
      /^.*\.avature\.net\/.+/i,
      /^.*avature\.net.*/i,
    ],
  },
  {
    provider: 'SUCCESSFACTORS',
    patterns: [
      /^.*\.successfactors\.com.*/i,
      /^.*successfactors\.com.*/i,
    ],
  },
  {
    provider: 'TALEO',
    patterns: [
      /^.*\.taleo\.net.*/i,
      /^.*taleo\.net.*/i,
    ],
  },
  {
    provider: 'ICIMS',
    patterns: [
      /^.*\.icims\.com.*/i,
      /^.*icims\.com.*/i,
    ],
  },
  {
    provider: 'PHENOM',
    patterns: [
      /^.*\.phenompeople\.com.*/i,
      /^.*phenompeople\.com.*/i,
    ],
  },
];

export function detectProvider(urlString: string): Provider {
  try {
    const url = new URL(urlString);
    const hostname = url.hostname;
    const pathname = url.pathname;
    const fullPath = hostname + pathname;

    for (const { provider, patterns } of PROVIDER_PATTERNS) {
      for (const pattern of patterns) {
        if (typeof pattern === 'function') {
          if (pattern(url)) {
            return provider;
          }
        } else if (pattern.test(fullPath)) {
          return provider;
        }
      }
    }

    return 'UNKNOWN';
  } catch (error) {
    return 'UNKNOWN';
  }
}

export function extractGreenhouseIds(urlString: string): { board: string; jobId: string } | null {
  const match = urlString.match(/job-boards\.greenhouse\.io\/([^/]+)\/jobs\/(\d+)/i) ||
                urlString.match(/boards\.greenhouse\.io\/([^/]+)\/jobs\/(\d+)/i);
  if (match) {
    return { board: match[1], jobId: match[2] };
  }
  return null;
}

export function extractLeverIds(urlString: string): { account: string; postingId: string } | null {
  const match = urlString.match(/jobs\.lever\.co\/([^/]+)\/([^/]+)/i);
  if (match) {
    return { account: match[1], postingId: match[2] };
  }
  return null;
}

export function extractAshbyIds(urlString: string): { company: string; jobId: string } | null {
  const match = urlString.match(/jobs\.ashbyhq\.com\/([^/]+)\/([^/]+)/i);
  if (match) {
    return { company: match[1], jobId: match[2] };
  }
  return null;
}

export function extractSmartRecruitersIds(urlString: string): { companyIdentifier: string; postingId: string } | null {
  const match = urlString.match(/jobs\.smartrecruiters\.com\/([^/]+)\/([^/]+)/i);
  if (match) {
    return { companyIdentifier: match[1], postingId: match[2] };
  }
  return null;
}
